Enable bare minimum of OMG on 32-bits

aoikonomopoulos opened the pull request:
#### 037cafb096cd36a13e2371bf911b5c849c2d2118
<pre>
Enable bare minimum of OMG on 32-bits
<a href="https://bugs.webkit.org/show_bug.cgi?id=273869">https://bugs.webkit.org/show_bug.cgi?id=273869</a>

Reviewed by Justin Michaud.

Add a minimal set of changes to get two simple testcases to run.

The main changes are:

- Copy WasmOMGIRGenerator.cpp to WasmOMGIRGenerator32_64.cpp.

- Copy B3LowerToAir.cpp to B3LowerToAir32_64.cpp.

- Change B3LowerToAir32_64.cpp to map B3 Values to either 1 or 2 Tmps.

- Introduce B3::ValueRep::RegisterPair and Air::Arg::TmpPair. We need this to
  represent Int64 values for argument passing/return. This is mainly used for
  Patchpoints, as the alternative of lowering Int64 values to 2 Air args would
  significantly complicate the stackmap code.

- Teach WasmCallingConvention.h to return JSValueRegs with the tagGPR set to
  InvalidGPRReg for Int32 values. This means we need to pass the valueSize all
  the way through to the register assignment point.

- Add 32-bit-specific lowering code in B3LowerToAir.cpp. For regular
  operations on Int64 values, emit code that either uses a special Air form or
  (in cases where that&apos;s possible) use the existing 32-bit forms and combine
  the results.

- Implement OpcodeOrigin. For simplicity&apos;s sake, use a 32-bit word, even if that
  truncates the offset to 24 bits.

Supporting changes:

- Disable parts of WasmOMGIRGenerator32_64.cpp we&apos;re not ready to enable yet (not required by
  our testcases).
- Make some minimal 32-bit adjustments to WasmOMGIRGenerator32_64.cpp to get
  the testcases to run.

With these changes, cc-int-to-int.js and i64-call.js tier up to OMG (with
JSC_useOMGJIT=true) and complete successfully.

* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/b3/B3ConstrainedValue.h:
(JSC::B3::ConstrainedValue::ConstrainedValue):
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/B3LowerToAir32_64.cpp: Copied from Source/JavaScriptCore/b3/B3LowerToAir.cpp.
(JSC::B3::lowerToAir):
* Source/JavaScriptCore/b3/B3StackmapSpecial.cpp:
(JSC::B3::StackmapSpecial::forEachArgImpl):
(JSC::B3::StackmapSpecial::isArgValidForType):
(JSC::B3::StackmapSpecial::isArgValidForRep):
(JSC::B3::StackmapSpecial::repForArg):
* Source/JavaScriptCore/b3/B3Validate.cpp:
* Source/JavaScriptCore/b3/B3ValueRep.cpp:
(JSC::B3::ValueRep::addUsedRegistersTo const):
(JSC::B3::ValueRep::dump const):
(WTF::printInternal):
* Source/JavaScriptCore/b3/B3ValueRep.h:
(JSC::B3::ValueRep::ValueRep):
(JSC::B3::ValueRep::regPair):
(JSC::B3::ValueRep::isRegPair const):
(JSC::B3::ValueRep::regLo const):
(JSC::B3::ValueRep::regHi const):
(JSC::B3::ValueRep::isGPRPair const):
* Source/JavaScriptCore/b3/air/AirArg.cpp:
(JSC::B3::Air::Arg::jsHash const):
(JSC::B3::Air::Arg::dump const):
(WTF::printInternal):
* Source/JavaScriptCore/b3/air/AirArg.h:
(JSC::B3::Air::Arg::Arg):
(JSC::B3::Air::Arg::isTmpPair const):
(JSC::B3::Air::Arg::tmpHi const):
(JSC::B3::Air::Arg::tmpLo const):
(JSC::B3::Air::Arg::isGPTmpPair const):
(JSC::B3::Air::Arg::regHi const):
(JSC::B3::Air::Arg::regLo const):
(JSC::B3::Air::Arg::isGP const):
(JSC::B3::Air::Arg::isFP const):
(JSC::B3::Air::Arg::hasBank const):
(JSC::B3::Air::Arg::isValidForm const):
(JSC::B3::Air::Arg::forEachTmpFast):
(JSC::B3::Air::Arg::forEachTmp):
* Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
(JSC::AssemblyHelpers::checkWasmStackOverflow):
* Source/JavaScriptCore/jit/AssemblyHelpers.h:
* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::emitEntryTierUpCheck):
* Source/JavaScriptCore/wasm/WasmCallingConvention.h:
(JSC::Wasm::ArgumentLocation::ArgumentLocation):
(JSC::Wasm::WasmCallingConvention::marshallRegs const):
(JSC::Wasm::WasmCallingConvention::marshallLocationImpl const):
(JSC::Wasm::WasmCallingConvention::marshallLocation const):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp: Copied from Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp.
(JSC::Wasm::OMGIRGenerator::ControlData::ControlData):
(JSC::Wasm::OMGIRGenerator::ControlData::isIf):
(JSC::Wasm::OMGIRGenerator::ControlData::isTry):
(JSC::Wasm::OMGIRGenerator::ControlData::isAnyCatch):
(JSC::Wasm::OMGIRGenerator::ControlData::isTopLevel):
(JSC::Wasm::OMGIRGenerator::ControlData::isLoop):
(JSC::Wasm::OMGIRGenerator::ControlData::isBlock):
(JSC::Wasm::OMGIRGenerator::ControlData::isCatch):
(JSC::Wasm::OMGIRGenerator::ControlData::dump const):
(JSC::Wasm::OMGIRGenerator::ControlData::blockType const):
(JSC::Wasm::OMGIRGenerator::ControlData::signature const):
(JSC::Wasm::OMGIRGenerator::ControlData::hasNonVoidresult const):
(JSC::Wasm::OMGIRGenerator::ControlData::targetBlockForBranch):
(JSC::Wasm::OMGIRGenerator::ControlData::convertIfToBlock):
(JSC::Wasm::OMGIRGenerator::ControlData::convertTryToCatch):
(JSC::Wasm::OMGIRGenerator::ControlData::convertTryToCatchAll):
(JSC::Wasm::OMGIRGenerator::ControlData::branchTargetArity const):
(JSC::Wasm::OMGIRGenerator::ControlData::branchTargetType const):
(JSC::Wasm::OMGIRGenerator::ControlData::tryStart const):
(JSC::Wasm::OMGIRGenerator::ControlData::tryEnd const):
(JSC::Wasm::OMGIRGenerator::ControlData::tryDepth const):
(JSC::Wasm::OMGIRGenerator::ControlData::catchKind const):
(JSC::Wasm::OMGIRGenerator::ControlData::exception const):
(JSC::Wasm::OMGIRGenerator::ControlData::stackSize const):
(JSC::Wasm::OMGIRGenerator::emptyExpression):
(JSC::Wasm::OMGIRGenerator::fail const):
(JSC::Wasm::OMGIRGenerator::advanceCallSiteIndex):
(JSC::Wasm::OMGIRGenerator::callSiteIndex const):
(JSC::Wasm::OMGIRGenerator::usesSIMD):
(JSC::Wasm::OMGIRGenerator::notifyFunctionUsesSIMD):
(JSC::Wasm::OMGIRGenerator::addConstant):
(JSC::Wasm::OMGIRGenerator::addExtractLane):
(JSC::Wasm::OMGIRGenerator::addReplaceLane):
(JSC::Wasm::OMGIRGenerator::addSIMDI_V):
(JSC::Wasm::OMGIRGenerator::addSIMDV_V):
(JSC::Wasm::OMGIRGenerator::addSIMDBitwiseSelect):
(JSC::Wasm::OMGIRGenerator::addSIMDRelOp):
(JSC::Wasm::OMGIRGenerator::fixupOutOfBoundsIndicesForSwizzle):
(JSC::Wasm::OMGIRGenerator::addSIMDV_VV):
(JSC::Wasm::OMGIRGenerator::addSIMDRelaxedFMA):
(JSC::Wasm::OMGIRGenerator::addEndToUnreachable):
(JSC::Wasm::OMGIRGenerator::endTopLevel):
(JSC::Wasm::OMGIRGenerator::setParser):
(JSC::Wasm::OMGIRGenerator::willParseOpcode):
(JSC::Wasm::OMGIRGenerator::didParseOpcode):
(JSC::Wasm::OMGIRGenerator::didFinishParsingLocals):
(JSC::Wasm::OMGIRGenerator::didPopValueFromStack):
(JSC::Wasm::OMGIRGenerator::getTypeDefinition):
(JSC::Wasm::OMGIRGenerator::addStackMap):
(JSC::Wasm::OMGIRGenerator::takeStackmaps):
(JSC::Wasm::OMGIRGenerator::takeExceptionHandlers):
(JSC::Wasm::OMGIRGenerator::emitPrepareWasmOperation):
(JSC::Wasm::OMGIRGenerator::callWasmOperation):
(JSC::Wasm::OMGIRGenerator::outerLoopIndex const):
(JSC::Wasm::OMGIRGenerator::getPushVariable):
(JSC::Wasm::OMGIRGenerator::push):
(JSC::Wasm::OMGIRGenerator::get):
(JSC::Wasm::OMGIRGenerator::set):
(JSC::Wasm::OMGIRGenerator::useSignalingMemory const):
(JSC::Wasm::OMGIRGenerator::instanceValue):
(JSC::Wasm::OMGIRGenerator::baseMemoryValue):
(JSC::Wasm::OMGIRGenerator::boundsCheckingSizeValue):
(JSC::Wasm::OMGIRGenerator::fixupPointerPlusOffset):
(JSC::Wasm::OMGIRGenerator::restoreWasmContextInstance):
(JSC::Wasm::OMGIRGenerator::computeStackCheckSize):
(JSC::Wasm::OMGIRGenerator::OMGIRGenerator):
(JSC::Wasm::OMGIRGenerator::restoreWebAssemblyGlobalState):
(JSC::Wasm::OMGIRGenerator::reloadMemoryRegistersFromInstance):
(JSC::Wasm::OMGIRGenerator::emitExceptionCheck):
(JSC::Wasm::OMGIRGenerator::constant):
(JSC::Wasm::OMGIRGenerator::framePointer):
(JSC::Wasm::OMGIRGenerator::insertEntrySwitch):
(JSC::Wasm::OMGIRGenerator::insertConstants):
(JSC::Wasm::OMGIRGenerator::toB3ResultType):
(JSC::Wasm::OMGIRGenerator::addLocal):
(JSC::Wasm::OMGIRGenerator::addDrop):
(JSC::Wasm::OMGIRGenerator::addInlinedArguments):
(JSC::Wasm::OMGIRGenerator::addArguments):
(JSC::Wasm::OMGIRGenerator::addRefIsNull):
(JSC::Wasm::OMGIRGenerator::addTableGet):
(JSC::Wasm::OMGIRGenerator::addTableSet):
(JSC::Wasm::OMGIRGenerator::addRefFunc):
(JSC::Wasm::OMGIRGenerator::addRefAsNonNull):
(JSC::Wasm::OMGIRGenerator::addRefEq):
(JSC::Wasm::OMGIRGenerator::addTableInit):
(JSC::Wasm::OMGIRGenerator::addElemDrop):
(JSC::Wasm::OMGIRGenerator::addTableSize):
(JSC::Wasm::OMGIRGenerator::addTableGrow):
(JSC::Wasm::OMGIRGenerator::addTableFill):
(JSC::Wasm::OMGIRGenerator::addTableCopy):
(JSC::Wasm::OMGIRGenerator::getLocal):
(JSC::Wasm::OMGIRGenerator::addUnreachable):
(JSC::Wasm::OMGIRGenerator::addCrash):
(JSC::Wasm::OMGIRGenerator::emitIndirectCall):
(JSC::Wasm::OMGIRGenerator::addGrowMemory):
(JSC::Wasm::OMGIRGenerator::addCurrentMemory):
(JSC::Wasm::OMGIRGenerator::addMemoryFill):
(JSC::Wasm::OMGIRGenerator::addMemoryInit):
(JSC::Wasm::OMGIRGenerator::addMemoryCopy):
(JSC::Wasm::OMGIRGenerator::addDataDrop):
(JSC::Wasm::OMGIRGenerator::traceValue):
(JSC::Wasm::OMGIRGenerator::traceCF):
(JSC::Wasm::OMGIRGenerator::setLocal):
(JSC::Wasm::OMGIRGenerator::getGlobal):
(JSC::Wasm::OMGIRGenerator::setGlobal):
(JSC::Wasm::OMGIRGenerator::emitWriteBarrierForJSWrapper):
(JSC::Wasm::OMGIRGenerator::emitWriteBarrier):
(JSC::Wasm::OMGIRGenerator::emitCheckAndPreparePointer):
(JSC::Wasm::sizeOfLoadOp):
(JSC::Wasm::OMGIRGenerator::memoryKind):
(JSC::Wasm::OMGIRGenerator::emitLoadOp):
(JSC::Wasm::OMGIRGenerator::load):
(JSC::Wasm::sizeOfStoreOp):
(JSC::Wasm::OMGIRGenerator::emitStoreOp):
(JSC::Wasm::OMGIRGenerator::store):
(JSC::Wasm::accessWidth):
(JSC::Wasm::sizeOfAtomicOpMemoryAccess):
(JSC::Wasm::OMGIRGenerator::sanitizeAtomicResult):
(JSC::Wasm::OMGIRGenerator::fixupPointerPlusOffsetForAtomicOps):
(JSC::Wasm::OMGIRGenerator::emitAtomicLoadOp):
(JSC::Wasm::OMGIRGenerator::atomicLoad):
(JSC::Wasm::OMGIRGenerator::emitAtomicStoreOp):
(JSC::Wasm::OMGIRGenerator::atomicStore):
(JSC::Wasm::OMGIRGenerator::emitAtomicBinaryRMWOp):
(JSC::Wasm::OMGIRGenerator::atomicBinaryRMW):
(JSC::Wasm::OMGIRGenerator::emitAtomicCompareExchange):
(JSC::Wasm::OMGIRGenerator::emitStructSet):
(JSC::Wasm::OMGIRGenerator::atomicCompareExchange):
(JSC::Wasm::OMGIRGenerator::atomicWait):
(JSC::Wasm::OMGIRGenerator::atomicNotify):
(JSC::Wasm::OMGIRGenerator::atomicFence):
(JSC::Wasm::OMGIRGenerator::truncSaturated):
(JSC::Wasm::OMGIRGenerator::addRefI31):
(JSC::Wasm::OMGIRGenerator::addI31GetS):
(JSC::Wasm::OMGIRGenerator::addI31GetU):
(JSC::Wasm::OMGIRGenerator::pushArrayNew):
(JSC::Wasm::OMGIRGenerator::getArrayTypeDefinition):
(JSC::Wasm::OMGIRGenerator::getArrayElementType):
(JSC::Wasm::OMGIRGenerator::getArrayRefType):
(JSC::Wasm::OMGIRGenerator::addArrayNew):
(JSC::Wasm::OMGIRGenerator::pushArrayNewFromSegment):
(JSC::Wasm::OMGIRGenerator::addArrayNewDefault):
(JSC::Wasm::OMGIRGenerator::addArrayNewData):
(JSC::Wasm::OMGIRGenerator::addArrayNewElem):
(JSC::Wasm::OMGIRGenerator::addArrayNewFixed):
(JSC::Wasm::OMGIRGenerator::addArrayGet):
(JSC::Wasm::OMGIRGenerator::emitArrayNullCheck):
(JSC::Wasm::OMGIRGenerator::emitArraySetUnchecked):
(JSC::Wasm::OMGIRGenerator::addArraySet):
(JSC::Wasm::OMGIRGenerator::addArrayLen):
(JSC::Wasm::OMGIRGenerator::addArrayFill):
(JSC::Wasm::OMGIRGenerator::addArrayCopy):
(JSC::Wasm::OMGIRGenerator::addArrayInitElem):
(JSC::Wasm::OMGIRGenerator::addArrayInitData):
(JSC::Wasm::OMGIRGenerator::addStructNew):
(JSC::Wasm::OMGIRGenerator::addStructNewDefault):
(JSC::Wasm::OMGIRGenerator::addStructGet):
(JSC::Wasm::OMGIRGenerator::addStructSet):
(JSC::Wasm::OMGIRGenerator::addRefTest):
(JSC::Wasm::OMGIRGenerator::addRefCast):
(JSC::Wasm::OMGIRGenerator::emitRefTestOrCast):
(JSC::Wasm::OMGIRGenerator::emitCheckOrBranchForCast):
(JSC::Wasm::OMGIRGenerator::emitLoadRTTFromFuncref):
(JSC::Wasm::OMGIRGenerator::emitLoadRTTFromObject):
(JSC::Wasm::OMGIRGenerator::emitNotRTTKind):
(JSC::Wasm::OMGIRGenerator::addAnyConvertExtern):
(JSC::Wasm::OMGIRGenerator::addExternConvertAny):
(JSC::Wasm::OMGIRGenerator::addSelect):
(JSC::Wasm::OMGIRGenerator::addSIMDSplat):
(JSC::Wasm::OMGIRGenerator::addSIMDShift):
(JSC::Wasm::OMGIRGenerator::addSIMDExtmul):
(JSC::Wasm::OMGIRGenerator::addSIMDShuffle):
(JSC::Wasm::OMGIRGenerator::addSIMDLoad):
(JSC::Wasm::OMGIRGenerator::addSIMDStore):
(JSC::Wasm::OMGIRGenerator::addSIMDLoadSplat):
(JSC::Wasm::OMGIRGenerator::addSIMDLoadLane):
(JSC::Wasm::OMGIRGenerator::addSIMDStoreLane):
(JSC::Wasm::OMGIRGenerator::addSIMDLoadExtend):
(JSC::Wasm::OMGIRGenerator::addSIMDLoadPad):
(JSC::Wasm::OMGIRGenerator::emitEntryTierUpCheck):
(JSC::Wasm::OMGIRGenerator::emitLoopTierUpCheck):
(JSC::Wasm::OMGIRGenerator::loadFromScratchBuffer):
(JSC::Wasm::OMGIRGenerator::connectControlAtEntrypoint):
(JSC::Wasm::OMGIRGenerator::addLoop):
(JSC::Wasm::OMGIRGenerator::addTopLevel):
(JSC::Wasm::OMGIRGenerator::addBlock):
(JSC::Wasm::OMGIRGenerator::addIf):
(JSC::Wasm::OMGIRGenerator::addElse):
(JSC::Wasm::OMGIRGenerator::addElseToUnreachable):
(JSC::Wasm::OMGIRGenerator::addTry):
(JSC::Wasm::OMGIRGenerator::addCatch):
(JSC::Wasm::OMGIRGenerator::preparePatchpointForExceptions):
(JSC::Wasm::OMGIRGenerator::addCatchToUnreachable):
(JSC::Wasm::OMGIRGenerator::addCatchAll):
(JSC::Wasm::OMGIRGenerator::addCatchAllToUnreachable):
(JSC::Wasm::OMGIRGenerator::emitCatchImpl):
(JSC::Wasm::OMGIRGenerator::addDelegate):
(JSC::Wasm::OMGIRGenerator::addDelegateToUnreachable):
(JSC::Wasm::OMGIRGenerator::addThrow):
(JSC::Wasm::OMGIRGenerator::addRethrow):
(JSC::Wasm::OMGIRGenerator::addInlinedReturn):
(JSC::Wasm::OMGIRGenerator::addReturn):
(JSC::Wasm::OMGIRGenerator::addBranch):
(JSC::Wasm::OMGIRGenerator::addBranchNull):
(JSC::Wasm::OMGIRGenerator::addBranchCast):
(JSC::Wasm::OMGIRGenerator::addSwitch):
(JSC::Wasm::OMGIRGenerator::endBlock):
(JSC::Wasm::OMGIRGenerator::createCallPatchpoint):
(JSC::Wasm::OMGIRGenerator::createTailCallPatchpoint):
(JSC::Wasm::OMGIRGenerator::canInline const):
(JSC::Wasm::OMGIRGenerator::emitInlineDirectCall):
(JSC::Wasm::OMGIRGenerator::addCall):
(JSC::Wasm::OMGIRGenerator::addCallIndirect):
(JSC::Wasm::OMGIRGenerator::addCallRef):
(JSC::Wasm::OMGIRGenerator::unify):
(JSC::Wasm::OMGIRGenerator::unifyValuesWithBlock):
(JSC::Wasm::dumpExpressionStack):
(JSC::Wasm::OMGIRGenerator::dump):
(JSC::Wasm::OMGIRGenerator::origin):
(JSC::Wasm::shouldDumpIRFor):
(JSC::Wasm::parseAndCompileOMG):
(JSC::Wasm::OMGIRGenerator::emitChecksForModOrDiv):
(JSC::Wasm::OMGIRGenerator::addI32DivS):
(JSC::Wasm::OMGIRGenerator::addI32RemS):
(JSC::Wasm::OMGIRGenerator::addI32DivU):
(JSC::Wasm::OMGIRGenerator::addI32RemU):
(JSC::Wasm::OMGIRGenerator::addI64DivS):
(JSC::Wasm::OMGIRGenerator::addI64RemS):
(JSC::Wasm::OMGIRGenerator::addI64DivU):
(JSC::Wasm::OMGIRGenerator::addI64RemU):
(JSC::Wasm::OMGIRGenerator::addI32Ctz):
(JSC::Wasm::OMGIRGenerator::addI64Ctz):
(JSC::Wasm::OMGIRGenerator::addI32Popcnt):
(JSC::Wasm::OMGIRGenerator::addI64Popcnt):
(JSC::Wasm::OMGIRGenerator::addF64ConvertUI64):
(JSC::Wasm::OMGIRGenerator::addF32ConvertUI64):
(JSC::Wasm::OMGIRGenerator::addF64Nearest):
(JSC::Wasm::OMGIRGenerator::addF32Nearest):
(JSC::Wasm::OMGIRGenerator::addF64Trunc):
(JSC::Wasm::OMGIRGenerator::addF32Trunc):
(JSC::Wasm::OMGIRGenerator::addI32TruncSF64):
(JSC::Wasm::OMGIRGenerator::addI32TruncSF32):
(JSC::Wasm::OMGIRGenerator::addI32TruncUF64):
(JSC::Wasm::OMGIRGenerator::addI32TruncUF32):
(JSC::Wasm::OMGIRGenerator::addI64TruncSF64):
(JSC::Wasm::OMGIRGenerator::addI64TruncUF64):
(JSC::Wasm::OMGIRGenerator::addI64TruncSF32):
(JSC::Wasm::OMGIRGenerator::addI64TruncUF32):
* Source/JavaScriptCore/wasm/WasmOpcodeOrigin.h:
(JSC::Wasm::OpcodeOrigin::OpcodeOrigin):
(JSC::Wasm::OpcodeOrigin::opcode const):
(JSC::Wasm::OpcodeOrigin::location const):
* Source/WTF/wtf/PlatformEnable.h:

Canonical link: <a href="https://commits.webkit.org/279741@main">https://commits.webkit.org/279741@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4acc2decfc1c5c76ea96a23c486884ff1977afda

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54364 "24 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33767 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6921 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57643 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5095 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41296 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/5043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44026 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3407 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56457 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/31939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47078 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25162 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/28747 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4402 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3238 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/47725 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/50511 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4611 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59234 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/53871 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/29583 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/5043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51450 "Found 1 new test failure: imported/w3c/web-platform-tests/pointerevents/pointerevent_pointer_boundary_events_after_reappending_last_over_target.html (failure)") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/30742 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47171 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50812 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11851 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/31716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/66170 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30532 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12604 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/wasm/WasmCallingConvention.h:
```diff
@@ -117,34 +135,55 @@ class WasmCallingConvention {
     WTF_MAKE_NONCOPYABLE(WasmCallingConvention);
 
 private:
+    template <typename RegType>
```
Can we just make a separate arm marshaller?

aoikonomopoulos replied:
Yah, that sounds simpler.

aoikonomopoulos replied:
To be clear, would you prefer a separate `WasmCallingConventionArmThumb2&` class (like we do for `CCallingConventionArmThumb2` in the same file) or a different (`CPU(ARM_THUMB2)`) implementation of the methods in the same class. The former would ensure maximal separation, but then we can't have the callers be architecture-agnostic (at least, not without defining an abstract class interface).

MenloDorian replied:
I'm not sure a separate `WasmCallingConvention` would be simpler.  It looks like there's a lot of code in common right now.

aoikonomopoulos replied:
Kk, will upload a new version which only makes minimal changes to this file, as in an earlier iteration.

justinmichaud commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -70,6 +70,57 @@ namespace JSC { namespace B3 {
+    Tmp lo, hi;
+};
+
+using SomeTmp = MaybeWide<Tmp, WideTmp>;
```
Maybe we can make Tmp->MachineTmp, and SomeTmp->LogicalTmp? Or something like that? Let's get a second opinion

aoikonomopoulos replied:
Yah,  `SomeTmp` is a bit of a placeholder name. Perhaps `SomeTmp` -> `TmpHolder`? To me `MachineTmp` implies there's something special about the `Tmp`, so renaming it feels like it could mislead the reader. More suggestions welcome.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp:
```diff
@@ -1097,11 +1094,12 @@ OMGIRGenerator::OMGIRGenerator(const ModuleInformation& info, OptimizingJITCalle
 
-    m_proc.pinRegister(GPRInfo::wasmBaseMemoryPointer);
     m_proc.pinRegister(GPRInfo::wasmContextInstancePointer);
+#if OMG_JSVALUE_32_64_PINNED_MEMORY_REGISTERS
```
Are we planning to keep these?

justinmichaud replied:
Ah, for the follow-up. I see

aoikonomopoulos replied:
To be clear, I would leave the PINNED_MEMORY_REGISTERS guard around until OMG is fully-functional. The other OMG_JSVALUE32_64 guards are temporary, to be removed by follow-up PRs that adjust more of `WasmOMGIRGenerator32_64.cpp`.

MenloDorian commented on Source/JavaScriptCore/b3/B3ValueRep.h:
```diff
@@ -128,6 +137,12 @@ class ValueRep {
+            u.registerPair.regHi = location.jsr().tagGPR();
+            u.registerPair.regLo = location.jsr().payloadGPR();
+            break;
+#endif
```
The following code looks unreachable on JSVALUE32_64.  Should this be an #if #else case?

aoikonomopoulos replied:
Sure.

justinmichaud reviewed (approved):
r=me

aoikonomopoulos committed:
Enable bare minimum of OMG on 32-bits
https://bugs.webkit.org/show_bug.cgi?id=273869

Reviewed by Justin Michaud.

Add a minimal set of changes to get two simple testcases to run.

The main changes are:

- Copy WasmOMGIRGenerator.cpp to WasmOMGIRGenerator32_64.cpp.

- Copy B3LowerToAir.cpp to B3LowerToAir32_64.cpp.

- Change B3LowerToAir32_64.cpp to map B3 Values to either 1 or 2 Tmps.

- Introduce B3::ValueRep::RegisterPair and Air::Arg::TmpPair. We need this to
  represent Int64 values for argument passing/return. This is mainly used for
  Patchpoints, as the alternative of lowering Int64 values to 2 Air args would
  significantly complicate the stackmap code.

- Teach WasmCallingConvention.h to return JSValueRegs with the tagGPR set to
  InvalidGPRReg for Int32 values. This means we need to pass the valueSize all
  the way through to the register assignment point.

- Add 32-bit-specific lowering code in B3LowerToAir.cpp. For regular
  operations on Int64 values, emit code that either uses a special Air form or
  (in cases where that's possible) use the existing 32-bit forms and combine
  the results.

- Implement OpcodeOrigin. For simplicity's sake, use a 32-bit word, even if that
  truncates the offset to 24 bits.

Supporting changes:

- Disable parts of WasmOMGIRGenerator32_64.cpp we're not ready to enable yet (not required by
  our testcases).
- Make some minimal 32-bit adjustments to WasmOMGIRGenerator32_64.cpp to get
  the testcases to run.

With these changes, cc-int-to-int.js and i64-call.js tier up to OMG (with
JSC_useOMGJIT=true) and complete successfully.

* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/b3/B3ConstrainedValue.h:
(JSC::B3::ConstrainedValue::ConstrainedValue):
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/B3LowerToAir32_64.cpp: Copied from Source/JavaScriptCore/b3/B3LowerToAir.cpp.
(JSC::B3::lowerToAir):
* Source/JavaScriptCore/b3/B3StackmapSpecial.cpp:
(JSC::B3::StackmapSpecial::forEachArgImpl):
(JSC::B3::StackmapSpecial::isArgValidForType):
(JSC::B3::StackmapSpecial::isArgValidForRep):
(JSC::B3::StackmapSpecial::repForArg):
* Source/JavaScriptCore/b3/B3Validate.cpp:
* Source/JavaScriptCore/b3/B3ValueRep.cpp:
(JSC::B3::ValueRep::addUsedRegistersTo const):
(JSC::B3::ValueRep::dump const):
(WTF::printInternal):
* Source/JavaScriptCore/b3/B3ValueRep.h:
(JSC::B3::ValueRep::ValueRep):
(JSC::B3::ValueRep::regPair):
(JSC::B3::ValueRep::isRegPair const):
(JSC::B3::ValueRep::regLo const):
(JSC::B3::ValueRep::regHi const):
(JSC::B3::ValueRep::isGPRPair const):
* Source/JavaScriptCore/b3/air/AirArg.cpp:
(JSC::B3::Air::Arg::jsHash const):
(JSC::B3::Air::Arg::dump const):
(WTF::printInternal):
* Source/JavaScriptCore/b3/air/AirArg.h:
(JSC::B3::Air::Arg::Arg):
(JSC::B3::Air::Arg::isTmpPair const):
(JSC::B3::Air::Arg::tmpHi const):
(JSC::B3::Air::Arg::tmpLo const):
(JSC::B3::Air::Arg::isGPTmpPair const):
(JSC::B3::Air::Arg::regHi const):
(JSC::B3::Air::Arg::regLo const):
(JSC::B3::Air::Arg::isGP const):
(JSC::B3::Air::Arg::isFP const):
(JSC::B3::Air::Arg::hasBank const):
(JSC::B3::Air::Arg::isValidForm const):
(JSC::B3::Air::Arg::forEachTmpFast):
(JSC::B3::Air::Arg::forEachTmp):
* Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
(JSC::AssemblyHelpers::checkWasmStackOverflow):
* Source/JavaScriptCore/jit/AssemblyHelpers.h:
* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::emitEntryTierUpCheck):
* Source/JavaScriptCore/wasm/WasmCallingConvention.h:
(JSC::Wasm::ArgumentLocation::ArgumentLocation):
(JSC::Wasm::WasmCallingConvention::marshallRegs const):
(JSC::Wasm::WasmCallingConvention::marshallLocationImpl const):
(JSC::Wasm::WasmCallingConvention::marshallLocation const):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp: Copied from Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp.
(JSC::Wasm::OMGIRGenerator::ControlData::ControlData):
(JSC::Wasm::OMGIRGenerator::ControlData::isIf):
(JSC::Wasm::OMGIRGenerator::ControlData::isTry):
(JSC::Wasm::OMGIRGenerator::ControlData::isAnyCatch):
(JSC::Wasm::OMGIRGenerator::ControlData::isTopLevel):
(JSC::Wasm::OMGIRGenerator::ControlData::isLoop):
(JSC::Wasm::OMGIRGenerator::ControlData::isBlock):
(JSC::Wasm::OMGIRGenerator::ControlData::isCatch):
(JSC::Wasm::OMGIRGenerator::ControlData::dump const):
(JSC::Wasm::OMGIRGenerator::ControlData::blockType const):
(JSC::Wasm::OMGIRGenerator::ControlData::signature const):
(JSC::Wasm::OMGIRGenerator::ControlData::hasNonVoidresult const):
(JSC::Wasm::OMGIRGenerator::ControlData::targetBlockForBranch):
(JSC::Wasm::OMGIRGenerator::ControlData::convertIfToBlock):
(JSC::Wasm::OMGIRGenerator::ControlData::convertTryToCatch):
(JSC::Wasm::OMGIRGenerator::ControlData::convertTryToCatchAll):
(JSC::Wasm::OMGIRGenerator::ControlData::branchTargetArity const):
(JSC::Wasm::OMGIRGenerator::ControlData::branchTargetType const):
(JSC::Wasm::OMGIRGenerator::ControlData::tryStart const):
(JSC::Wasm::OMGIRGenerator::ControlData::tryEnd const):
(JSC::Wasm::OMGIRGenerator::ControlData::tryDepth const):
(JSC::Wasm::OMGIRGenerator::ControlData::catchKind const):
(JSC::Wasm::OMGIRGenerator::ControlData::exception const):
(JSC::Wasm::OMGIRGenerator::ControlData::stackSize const):
(JSC::Wasm::OMGIRGenerator::emptyExpression):
(JSC::Wasm::OMGIRGenerator::fail const):
(JSC::Wasm::OMGIRGenerator::advanceCallSiteIndex):
(JSC::Wasm::OMGIRGenerator::callSiteIndex const):
(JSC::Wasm::OMGIRGenerator::usesSIMD):
(JSC::Wasm::OMGIRGenerator::notifyFunctionUsesSIMD):
(JSC::Wasm::OMGIRGenerator::addConstant):
(JSC::Wasm::OMGIRGenerator::addExtractLane):
(JSC::Wasm::OMGIRGenerator::addReplaceLane):
(JSC::Wasm::OMGIRGenerator::addSIMDI_V):
(JSC::Wasm::OMGIRGenerator::addSIMDV_V):
(JSC::Wasm::OMGIRGenerator::addSIMDBitwiseSelect):
(JSC::Wasm::OMGIRGenerator::addSIMDRelOp):
(JSC::Wasm::OMGIRGenerator::fixupOutOfBoundsIndicesForSwizzle):
(JSC::Wasm::OMGIRGenerator::addSIMDV_VV):
(JSC::Wasm::OMGIRGenerator::addSIMDRelaxedFMA):
(JSC::Wasm::OMGIRGenerator::addEndToUnreachable):
(JSC::Wasm::OMGIRGenerator::endTopLevel):
(JSC::Wasm::OMGIRGenerator::setParser):
(JSC::Wasm::OMGIRGenerator::willParseOpcode):
(JSC::Wasm::OMGIRGenerator::didParseOpcode):
(JSC::Wasm::OMGIRGenerator::didFinishParsingLocals):
(JSC::Wasm::OMGIRGenerator::didPopValueFromStack):
(JSC::Wasm::OMGIRGenerator::getTypeDefinition):
(JSC::Wasm::OMGIRGenerator::addStackMap):
(JSC::Wasm::OMGIRGenerator::takeStackmaps):
(JSC::Wasm::OMGIRGenerator::takeExceptionHandlers):
(JSC::Wasm::OMGIRGenerator::emitPrepareWasmOperation):
(JSC::Wasm::OMGIRGenerator::callWasmOperation):
(JSC::Wasm::OMGIRGenerator::outerLoopIndex const):
(JSC::Wasm::OMGIRGenerator::getPushVariable):
(JSC::Wasm::OMGIRGenerator::push):
(JSC::Wasm::OMGIRGenerator::get):
(JSC::Wasm::OMGIRGenerator::set):
(JSC::Wasm::OMGIRGenerator::useSignalingMemory const):
(JSC::Wasm::OMGIRGenerator::instanceValue):
(JSC::Wasm::OMGIRGenerator::baseMemoryValue):
(JSC::Wasm::OMGIRGenerator::boundsCheckingSizeValue):
(JSC::Wasm::OMGIRGenerator::fixupPointerPlusOffset):
(JSC::Wasm::OMGIRGenerator::restoreWasmContextInstance):
(JSC::Wasm::OMGIRGenerator::computeStackCheckSize):
(JSC::Wasm::OMGIRGenerator::OMGIRGenerator):
(JSC::Wasm::OMGIRGenerator::restoreWebAssemblyGlobalState):
(JSC::Wasm::OMGIRGenerator::reloadMemoryRegistersFromInstance):
(JSC::Wasm::OMGIRGenerator::emitExceptionCheck):
(JSC::Wasm::OMGIRGenerator::constant):
(JSC::Wasm::OMGIRGenerator::framePointer):
(JSC::Wasm::OMGIRGenerator::insertEntrySwitch):
(JSC::Wasm::OMGIRGenerator::insertConstants):
(JSC::Wasm::OMGIRGenerator::toB3ResultType):
(JSC::Wasm::OMGIRGenerator::addLocal):
(JSC::Wasm::OMGIRGenerator::addDrop):
(JSC::Wasm::OMGIRGenerator::addInlinedArguments):
(JSC::Wasm::OMGIRGenerator::addArguments):
(JSC::Wasm::OMGIRGenerator::addRefIsNull):
(JSC::Wasm::OMGIRGenerator::addTableGet):
(JSC::Wasm::OMGIRGenerator::addTableSet):
(JSC::Wasm::OMGIRGenerator::addRefFunc):
(JSC::Wasm::OMGIRGenerator::addRefAsNonNull):
(JSC::Wasm::OMGIRGenerator::addRefEq):
(JSC::Wasm::OMGIRGenerator::addTableInit):
(JSC::Wasm::OMGIRGenerator::addElemDrop):
(JSC::Wasm::OMGIRGenerator::addTableSize):
(JSC::Wasm::OMGIRGenerator::addTableGrow):
(JSC::Wasm::OMGIRGenerator::addTableFill):
(JSC::Wasm::OMGIRGenerator::addTableCopy):
(JSC::Wasm::OMGIRGenerator::getLocal):
(JSC::Wasm::OMGIRGenerator::addUnreachable):
(JSC::Wasm::OMGIRGenerator::addCrash):
(JSC::Wasm::OMGIRGenerator::emitIndirectCall):
(JSC::Wasm::OMGIRGenerator::addGrowMemory):
(JSC::Wasm::OMGIRGenerator::addCurrentMemory):
(JSC::Wasm::OMGIRGenerator::addMemoryFill):
(JSC::Wasm::OMGIRGenerator::addMemoryInit):
(JSC::Wasm::OMGIRGenerator::addMemoryCopy):
(JSC::Wasm::OMGIRGenerator::addDataDrop):
(JSC::Wasm::OMGIRGenerator::traceValue):
(JSC::Wasm::OMGIRGenerator::traceCF):
(JSC::Wasm::OMGIRGenerator::setLocal):
(JSC::Wasm::OMGIRGenerator::getGlobal):
(JSC::Wasm::OMGIRGenerator::setGlobal):
(JSC::Wasm::OMGIRGenerator::emitWriteBarrierForJSWrapper):
(JSC::Wasm::OMGIRGenerator::emitWriteBarrier):
(JSC::Wasm::OMGIRGenerator::emitCheckAndPreparePointer):
(JSC::Wasm::sizeOfLoadOp):
(JSC::Wasm::OMGIRGenerator::memoryKind):
(JSC::Wasm::OMGIRGenerator::emitLoadOp):
(JSC::Wasm::OMGIRGenerator::load):
(JSC::Wasm::sizeOfStoreOp):
(JSC::Wasm::OMGIRGenerator::emitStoreOp):
(JSC::Wasm::OMGIRGenerator::store):
(JSC::Wasm::accessWidth):
(JSC::Wasm::sizeOfAtomicOpMemoryAccess):
(JSC::Wasm::OMGIRGenerator::sanitizeAtomicResult):
(JSC::Wasm::OMGIRGenerator::fixupPointerPlusOffsetForAtomicOps):
(JSC::Wasm::OMGIRGenerator::emitAtomicLoadOp):
(JSC::Wasm::OMGIRGenerator::atomicLoad):
(JSC::Wasm::OMGIRGenerator::emitAtomicStoreOp):
(JSC::Wasm::OMGIRGenerator::atomicStore):
(JSC::Wasm::OMGIRGenerator::emitAtomicBinaryRMWOp):
(JSC::Wasm::OMGIRGenerator::atomicBinaryRMW):
(JSC::Wasm::OMGIRGenerator::emitAtomicCompareExchange):
(JSC::Wasm::OMGIRGenerator::emitStructSet):
(JSC::Wasm::OMGIRGenerator::atomicCompareExchange):
(JSC::Wasm::OMGIRGenerator::atomicWait):
(JSC::Wasm::OMGIRGenerator::atomicNotify):
(JSC::Wasm::OMGIRGenerator::atomicFence):
(JSC::Wasm::OMGIRGenerator::truncSaturated):
(JSC::Wasm::OMGIRGenerator::addRefI31):
(JSC::Wasm::OMGIRGenerator::addI31GetS):
(JSC::Wasm::OMGIRGenerator::addI31GetU):
(JSC::Wasm::OMGIRGenerator::pushArrayNew):
(JSC::Wasm::OMGIRGenerator::getArrayTypeDefinition):
(JSC::Wasm::OMGIRGenerator::getArrayElementType):
(JSC::Wasm::OMGIRGenerator::getArrayRefType):
(JSC::Wasm::OMGIRGenerator::addArrayNew):
(JSC::Wasm::OMGIRGenerator::pushArrayNewFromSegment):
(JSC::Wasm::OMGIRGenerator::addArrayNewDefault):
(JSC::Wasm::OMGIRGenerator::addArrayNewData):
(JSC::Wasm::OMGIRGenerator::addArrayNewElem):
(JSC::Wasm::OMGIRGenerator::addArrayNewFixed):
(JSC::Wasm::OMGIRGenerator::addArrayGet):
(JSC::Wasm::OMGIRGenerator::emitArrayNullCheck):
(JSC::Wasm::OMGIRGenerator::emitArraySetUnchecked):
(JSC::Wasm::OMGIRGenerator::addArraySet):
(JSC::Wasm::OMGIRGenerator::addArrayLen):
(JSC::Wasm::OMGIRGenerator::addArrayFill):
(JSC::Wasm::OMGIRGenerator::addArrayCopy):
(JSC::Wasm::OMGIRGenerator::addArrayInitElem):
(JSC::Wasm::OMGIRGenerator::addArrayInitData):
(JSC::Wasm::OMGIRGenerator::addStructNew):
(JSC::Wasm::OMGIRGenerator::addStructNewDefault):
(JSC::Wasm::OMGIRGenerator::addStructGet):
(JSC::Wasm::OMGIRGenerator::addStructSet):
(JSC::Wasm::OMGIRGenerator::addRefTest):
(JSC::Wasm::OMGIRGenerator::addRefCast):
(JSC::Wasm::OMGIRGenerator::emitRefTestOrCast):
(JSC::Wasm::OMGIRGenerator::emitCheckOrBranchForCast):
(JSC::Wasm::OMGIRGenerator::emitLoadRTTFromFuncref):
(JSC::Wasm::OMGIRGenerator::emitLoadRTTFromObject):
(JSC::Wasm::OMGIRGenerator::emitNotRTTKind):
(JSC::Wasm::OMGIRGenerator::addAnyConvertExtern):
(JSC::Wasm::OMGIRGenerator::addExternConvertAny):
(JSC::Wasm::OMGIRGenerator::addSelect):
(JSC::Wasm::OMGIRGenerator::addSIMDSplat):
(JSC::Wasm::OMGIRGenerator::addSIMDShift):
(JSC::Wasm::OMGIRGenerator::addSIMDExtmul):
(JSC::Wasm::OMGIRGenerator::addSIMDShuffle):
(JSC::Wasm::OMGIRGenerator::addSIMDLoad):
(JSC::Wasm::OMGIRGenerator::addSIMDStore):
(JSC::Wasm::OMGIRGenerator::addSIMDLoadSplat):
(JSC::Wasm::OMGIRGenerator::addSIMDLoadLane):
(JSC::Wasm::OMGIRGenerator::addSIMDStoreLane):
(JSC::Wasm::OMGIRGenerator::addSIMDLoadExtend):
(JSC::Wasm::OMGIRGenerator::addSIMDLoadPad):
(JSC::Wasm::OMGIRGenerator::emitEntryTierUpCheck):
(JSC::Wasm::OMGIRGenerator::emitLoopTierUpCheck):
(JSC::Wasm::OMGIRGenerator::loadFromScratchBuffer):
(JSC::Wasm::OMGIRGenerator::connectControlAtEntrypoint):
(JSC::Wasm::OMGIRGenerator::addLoop):
(JSC::Wasm::OMGIRGenerator::addTopLevel):
(JSC::Wasm::OMGIRGenerator::addBlock):
(JSC::Wasm::OMGIRGenerator::addIf):
(JSC::Wasm::OMGIRGenerator::addElse):
(JSC::Wasm::OMGIRGenerator::addElseToUnreachable):
(JSC::Wasm::OMGIRGenerator::addTry):
(JSC::Wasm::OMGIRGenerator::addCatch):
(JSC::Wasm::OMGIRGenerator::preparePatchpointForExceptions):
(JSC::Wasm::OMGIRGenerator::addCatchToUnreachable):
(JSC::Wasm::OMGIRGenerator::addCatchAll):
(JSC::Wasm::OMGIRGenerator::addCatchAllToUnreachable):
(JSC::Wasm::OMGIRGenerator::emitCatchImpl):
(JSC::Wasm::OMGIRGenerator::addDelegate):
(JSC::Wasm::OMGIRGenerator::addDelegateToUnreachable):
(JSC::Wasm::OMGIRGenerator::addThrow):
(JSC::Wasm::OMGIRGenerator::addRethrow):
(JSC::Wasm::OMGIRGenerator::addInlinedReturn):
(JSC::Wasm::OMGIRGenerator::addReturn):
(JSC::Wasm::OMGIRGenerator::addBranch):
(JSC::Wasm::OMGIRGenerator::addBranchNull):
(JSC::Wasm::OMGIRGenerator::addBranchCast):
(JSC::Wasm::OMGIRGenerator::addSwitch):
(JSC::Wasm::OMGIRGenerator::endBlock):
(JSC::Wasm::OMGIRGenerator::createCallPatchpoint):
(JSC::Wasm::OMGIRGenerator::createTailCallPatchpoint):
(JSC::Wasm::OMGIRGenerator::canInline const):
(JSC::Wasm::OMGIRGenerator::emitInlineDirectCall):
(JSC::Wasm::OMGIRGenerator::addCall):
(JSC::Wasm::OMGIRGenerator::addCallIndirect):
(JSC::Wasm::OMGIRGenerator::addCallRef):
(JSC::Wasm::OMGIRGenerator::unify):
(JSC::Wasm::OMGIRGenerator::unifyValuesWithBlock):
(JSC::Wasm::dumpExpressionStack):
(JSC::Wasm::OMGIRGenerator::dump):
(JSC::Wasm::OMGIRGenerator::origin):
(JSC::Wasm::shouldDumpIRFor):
(JSC::Wasm::parseAndCompileOMG):
(JSC::Wasm::OMGIRGenerator::emitChecksForModOrDiv):
(JSC::Wasm::OMGIRGenerator::addI32DivS):
(JSC::Wasm::OMGIRGenerator::addI32RemS):
(JSC::Wasm::OMGIRGenerator::addI32DivU):
(JSC::Wasm::OMGIRGenerator::addI32RemU):
(JSC::Wasm::OMGIRGenerator::addI64DivS):
(JSC::Wasm::OMGIRGenerator::addI64RemS):
(JSC::Wasm::OMGIRGenerator::addI64DivU):
(JSC::Wasm::OMGIRGenerator::addI64RemU):
(JSC::Wasm::OMGIRGenerator::addI32Ctz):
(JSC::Wasm::OMGIRGenerator::addI64Ctz):
(JSC::Wasm::OMGIRGenerator::addI32Popcnt):
(JSC::Wasm::OMGIRGenerator::addI64Popcnt):
(JSC::Wasm::OMGIRGenerator::addF64ConvertUI64):
(JSC::Wasm::OMGIRGenerator::addF32ConvertUI64):
(JSC::Wasm::OMGIRGenerator::addF64Nearest):
(JSC::Wasm::OMGIRGenerator::addF32Nearest):
(JSC::Wasm::OMGIRGenerator::addF64Trunc):
(JSC::Wasm::OMGIRGenerator::addF32Trunc):
(JSC::Wasm::OMGIRGenerator::addI32TruncSF64):
(JSC::Wasm::OMGIRGenerator::addI32TruncSF32):
(JSC::Wasm::OMGIRGenerator::addI32TruncUF64):
(JSC::Wasm::OMGIRGenerator::addI32TruncUF32):
(JSC::Wasm::OMGIRGenerator::addI64TruncSF64):
(JSC::Wasm::OMGIRGenerator::addI64TruncUF64):
(JSC::Wasm::OMGIRGenerator::addI64TruncSF32):
(JSC::Wasm::OMGIRGenerator::addI64TruncUF32):
* Source/JavaScriptCore/wasm/WasmOpcodeOrigin.h:
(JSC::Wasm::OpcodeOrigin::OpcodeOrigin):
(JSC::Wasm::OpcodeOrigin::opcode const):
(JSC::Wasm::OpcodeOrigin::location const):
* Source/WTF/wtf/PlatformEnable.h:

Canonical link: https://commits.webkit.org/279741@main
